ast/compile.go

```diff
@@ -757,6 +757,41 @@ func (c *Compiler) PassesTypeCheck(body Body) bool {
 	return len(errs) == 0
 }
 
+// PassesTypeCheckRules determines whether the given rules passes type checking
+func (c *Compiler) PassesTypeCheckRules(rules []*Rule) Errors {
+	elems := []util.T{}
+
+	for _, rule := range rules {
+		elems = append(elems, rule)
+	}
+
+	// Load the global input schema if one was provided.
+	if c.schemaSet != nil {
+		if schema := c.schemaSet.Get(SchemaRootRef); schema != nil {
+
+			var allowNet []string
+			if c.capabilities != nil {
+				allowNet = c.capabilities.AllowNet
+			}
+
+			tpe, err := loadSchema(schema, allowNet)
+			if err != nil {
+				return Errors{NewError(TypeErr, nil, err.Error())}
+			}
+			c.inputType = tpe
+		}
+	}
+
+	var as *AnnotationSet
+	if c.useTypeCheckAnnotations {
+		as = c.annotationSet
+	}
+
+	checker := newTypeChecker().WithSchemaSet(c.schemaSet).WithInputType(c.inputType)
+	_, errs := checker.CheckTypes(c.TypeEnv, elems, as)
```

Comment:
I believe `CheckTypes` expects the passed rules to be sorted by how they internally depend on each other. Is that a requirement here, too? And if so, are they?

Reply:
At this point we're passing the rule and its dependencies without any sorting. The tests seem to work but I'll look into cases where there could be issues.

Reply:
The code here adds the rule and its dependencies https://github.com/open-policy-agent/opa/pull/6141/files#diff-f7251a508a59c5ce8cbffcb6a838a9b5e17471c2766707bb882447d18d522737R40. That seems to be enough. WDYT?